framework/src/actions/CreateProblemAction.C

```diff
@@ -81,6 +104,13 @@ CreateProblemAction::act()
           _factory.create<FEProblemBase>(_type, getParam<std::string>("name"), _moose_object_pars);
       if (!_problem.get())
         mooseError("Problem has to be of a FEProblemBase type");
+
+      // if users provide a problem type, the type has to be an EigenProblem or its derived subclass
+      // when uing an eigen executioner
+      if (_app.useEigenExecutioner() && _type != "EigenProblem" &&
+          !(std::dynamic_pointer_cast<EigenProblem>(_problem).get()))
```

Comment:
Note that you don't have to use `.get()` to test for `nullptr`. You can test the smart pointer directly. This just cleans up the code. You do this elsewhere in this PR as well.